org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/semantictokens/SemanticTokensService.java

```diff
@@ -0,0 +1,224 @@
+					data.add(deltaLine == 0 ? deltaOffset : position.getCharacter()); // delta start relative to previous
+																						// token if on the same line or to 0
+					lastOffset = lightweihtPosition.getOffset();
+					data.add(lightweihtPosition.getLength()); // length
+					data.add(tokenType); // token type
+					data.add(0); // token modifiers
+				}
+			}
+	    } catch (IOException e) {
+	        throw new WrappedException(e);
+	    }
+		return new SemanticTokens(data); // resultId not set, no delta support
+	}
+
+	/**
+	 * Semantic tokens range.
+	 *
+	 * @param params
+	 *            the params
+	 * @param cancelIndicator
+	 *            the cancel indicator
+	 * @return the semantic tokens
+	 */
+	public SemanticTokens semanticTokensRange(final Document document, final XtextResource resource,
+			final SemanticTokensRangeParams params, final CancelIndicator cancelIndicator) {
+	    if (document == null || resource == null) {
+	        return new SemanticTokens(Collections.emptyList());
+	    }
+	    List<Integer> data = new ArrayList<>();
+		Position rangeStart = params.getRange().getStart();
+		Position rangeEnd = params.getRange().getEnd();
+		int lastOffset = 0;
+		int lastLine = 0;
+		for (LightweightPosition lightweihtPosition : getPositions(resource, params.getTextDocument(),
+				cancelIndicator)) {
+			// TODO: it would be nicer to filter before calculating the list of LightweightPosition, to make the
+			// computation faster
+			// TODO: use PositionReader
+			Position position = document.getPosition(lightweihtPosition.getOffset());
+			if (rangeStart.getLine() > position.getLine() && position.getCharacter() <= rangeStart.getCharacter()) {
+				continue;
+			}
+			if (rangeEnd.getLine() < position.getLine()) {
+				break;
+			}
+			Integer tokenType = getTokenType(lightweihtPosition.getIds());
+			if (tokenType != -1) {
+				int deltaOffset = lightweihtPosition.getOffset() - lastOffset;
+				int deltaLine = position.getLine() - lastLine;
+				data.add(deltaLine); // delta line
+				lastLine = position.getLine();
+				data.add(deltaLine == 0 ? deltaOffset : position.getCharacter()); // delta start relative to previous
+																					// token if on the same line or to 0
+				lastOffset = lightweihtPosition.getOffset();
+				data.add(lightweihtPosition.getLength()); // length
+				data.add(tokenType); // token type
+				data.add(0); // token modifiers
+			}
+		}
+		return new SemanticTokens(data); // resultId not set, no delta support
+	}
+
+	public void setTokenTypes(final List<String> tokenTypes) {
+		this.tokenTypes = tokenTypes;
+	}
+
+	public List<String> getTokenTypes() {
+		return this.tokenTypes;
+	}
+
+	private int getTokenType(final String[] tokenTypes) {
+		for (String tokenType : tokenTypes) {
+			Integer index = this.tokenTypes.indexOf(tokenType);
+			if (index != -1) {
+				return index;
+			}
+		}
+		return -1;
+	}
+
+	protected URI getURI(final TextDocumentIdentifier documentItem) {
```

Comment:
This method and the ops below look as if they should rather be reused from somewhere else. Is this the only place where a service needs to be found based on a TextDocumentIdentifier?

Reply:
I have removed the method, I think it was not needed, because we can take the URI of the Resource.